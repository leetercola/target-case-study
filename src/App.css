.App {
  text-align: center; }

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 80px; }

.App-header {
  background-color: #222;
  height: 150px;
  padding: 20px;
  color: white; }

.App-title {
  font-size: 1.5em; }

.App-intro {
  font-size: large; }

.catalog-entry-view {
  top: 40px;
  position: relative; }

.cta-buttons {
  text-align: center;
  margin: 0 auto;
  margin-top: 20px;
  margin-bottom: 20px; }

.carousel-wrapper {
  margin-bottom: 30px; }

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg); }
  to {
    transform: rotate(360deg); } }
