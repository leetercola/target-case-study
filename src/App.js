import React, { Component } from 'react';
import './App.css';
import {Container, Row, Col, InputGroup, InputGroupAddon, InputGroupText, Input, Button} from 'reactstrap';
import _ from 'lodash';

import CaItemCarousel from './components/catalog-entry-item-carousel';
import Reviews from './components/reviews';

class App extends Component {
    render() {
        const {
            ItemData,
        } = this.props;

        const {
            CustomerReview,
            Offers:[firstOffer],
            Promotions,
            inventoryStatus,
            ItemDescription,
            ReturnPolicy,
            purchasingChannelCode,
        } = ItemData;   

        const {
            formattedPriceValue,
            priceQualifier
        } = firstOffer.OfferPrice[0];

        // some inline functions and element creation for brevity but these would ideally be broken into seperate functions or components for composition and testability.
        const promoElements = Promotions.map((item) => <p key={_.uniqueId('promo-elements')}className="text-danger font-weight-light">{item.Description[0].shortDescription}</p>);

        const showAddToCart = inventoryStatus === 'Online' && (purchasingChannelCode === '0' || purchasingChannelCode === '1');

        const showPickUpInStore = inventoryStatus === 'InStore' && (purchasingChannelCode === '0' || purchasingChannelCode === '2');
        console.log(showPickUpInStore);

        return (
            <Container className='catalog-entry-view'>
                <Row>
                    <Col sm={{size: 12}} md={{size: 6, order: 1}} >
                        <div className="clearfix carousel-wrapper">
                            <CaItemCarousel {...ItemData} />
                        </div>
                    </Col>
                    <Col sm={{size: 12}} md={{size: 6, order: 2}} className="item-description-card clearfix">
                        <p><strong>{formattedPriceValue}</strong> <small>{priceQualifier}</small></p>
                        <hr />
                        <div>
                            {promoElements}
                        </div>
                        <hr />
                        <InputGroup>
                            <InputGroupAddon addonType="prepend">
                                <Button>+</Button>
                            </InputGroupAddon>
                            <InputGroupAddon addonType="prepend">
                                <InputGroupText>
                                    quantity: 
                                </InputGroupText>
                            </InputGroupAddon>
                            <Input type="number" step="1" />
                            <InputGroupAddon addonType="append">
                                <Button>-</Button>
                            </InputGroupAddon>
                        </InputGroup>
                        <div className="form-group cta-buttons row">
                            {showPickUpInStore && (
                                <Col xs="6">
                                    <Button id="pickup_in_store" inverted block style={{background:'black'}}>Pick UP IN STORE</Button>
                                </Col>
                            )}
                            {showAddToCart && (
                                <Col xs="6">
                                    <Button id="add_to_cart" color="danger" block className="col-xs-6" style={{right:'0px', position: 'relative'}}>ADD TO CART</Button>
                                </Col>
                            )}
                        </div>
                        <div>
                            Returns | {<p dangerouslySetInnerHTML={{__html:ReturnPolicy[0].legalCopy}}/>}
                        </div>
                        <div>
                            <h3>Product highlights</h3>
                            <ul>
                                {
                                    //This is super bad and markup shouldn't exist in the data
                                    ItemDescription[0].features.map((item) => <li key={_.uniqueId('product-highlights-')}dangerouslySetInnerHTML={{__html:item}} />)
                                }
                            </ul>
                        </div>
                    </Col>
                    <Col sm={{size: 12}} md={{size: 6, order: 2}} className="reviews">
                        <Reviews {...CustomerReview[0]}/>
                    </Col>
                </Row>
            </Container>
        );
    }
}



export default App;