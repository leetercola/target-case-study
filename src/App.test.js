import React from 'react';
import App from './App';
import ItemData from './item-data.json';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';

afterEach = () => {
  shallow.unmount();
};

xit('renders add to cart if available online and purchasingChannelCode is 0 or 1 and does not show pick up in store', () => {
  const baseData = ItemData.CatalogEntryView[0];
  const output = shallow(<App ItemData={baseData} />);
  expect(output.find('#add_to_cart')).toBePresent();
  expect(output.find('#pickup_in_store')).not.toBePresent();
  expect(baseData.purchasingChannelCode === 1||2).toBeTruthy();
  expect(baseData.inventoryStatus).toEqual('Online');
  expect(output.text()).toMatchSnapshot();
  output.unmount();
});

it('renders pick up to store if inventoryStatus is InStore and purchasingChannelCode is 0 or 2', () => {
  let baseData = ItemData.CatalogEntryView[0];
  baseData.inventoryStatus = 'InStore';
  baseData.purchasingChannelCode = '0';
  const output = shallow(<App ItemData={baseData} />);
  expect(output.find('#pickup_in_store')).toBePresent();
  expect(output.find('#add_to_cart')).not.toBePresent();
  expect(baseData.purchasingChannelCode === 1||2).toBeTruthy();
  expect(baseData.inventoryStatus).toEqual('InStore');
  expect(output.text()).toMatchSnapshot();
  output.unmount();
})
