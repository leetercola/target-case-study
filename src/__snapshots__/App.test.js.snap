// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders add to cart if available online and purchasingChannelCode is 0 or 1 and does not show pick up in store 1`] = `"<Container />"`;

exports[`renders pick up to store if inventoryStatus is InStore and purchasingChannelCode is 0 or 2 1`] = `"<Container />"`;
